demos/golang/grpc_pingpong/prepare_ping_pong.sh

```diff
@@ -0,0 +1,67 @@
+#!/bin/bash
+set -e
+
+BLUE='\033[1;34m'
+NC='\033[0m'
+
+# sanity check
+FILE_SET="go.mod
+ping
+pong
+occlum_ping
+occlum_pong
+go.mod
+go.sum
+pingpong/pingpong.pb.go
+pingpong/pingpong_grpc.pb.go"
+for CURR_FILE in $FILE_SET
+do
+        if [ -f "$CURR_FILE" ]; then
+                rm $CURR_FILE
+        fi
+done
+
+DIR_SET="occlum_ping_instance
+occlum_pong_instance"
+for CURR_DIR in $DIR_SET
+do
+        if [ -d "$CURR_DIR" ]; then
+                rm -fr $CURR_DIR
+        fi
+done
+
+# enable Go modules for package management 
+export GO111MODULE=on
+
+# update PATH so that the protoc compiler can find the plugin:
+export PATH="$PATH:$(go env GOPATH)/bin"
+
+# assume that protoc is installed
+if ! type "protoc" > /dev/null; then
+        echo "Please install protoc"
+        exit 1
+fi
+
+# install protoc-gen-go and protoc-gen-go-grpc plugin
+if ! type "protoc-gen-go" > /dev/null; then
+        go get google.golang.org/protobuf/cmd/protoc-gen-go
+fi
+if ! type "protoc-gen-go-grpc" > /dev/null; then
+        go get google.golang.org/grpc/cmd/protoc-gen-go-grpc
+fi
+
+# compiling pingpong gRPC .proto file
+protoc --go-grpc_out=. --go_out=. pingpong/pingpong.proto
+
+# initialize Go modules for package management 
+go mod init grpc_pingpong
+
+# build pong image
+go build pong.go
```

Comment:
Somehow I got this following error message.
```
pingpong/pingpong_grpc.pb.go:14:11: undefined: grpc.SupportPackageIsVersion7
pingpong/pingpong_grpc.pb.go:71:39: undefined: grpc.ServiceRegistrar
```
Any idea?

Reply:
Guess it is relevant to recent commit to protoc-gen-go-grpc ("https://github.com/grpc/grpc-go/pull/3828"). Made changes to both scripts and the server code.

Unit-tested the changes. Seems fine now. Here are the outputs:

server side:
```
root@07ae377ab8cf:~/demos/golang/grpc_pingpong# ./run_pong_on_occlum.sh
/root/demos/golang/grpc_pingpong/occlum_pong_instance initialized as an Occlum instance
Enclave sign-tool: /opt/intel/sgxsdk/bin/x64/sgx_sign
Enclave sign-key: /opt/occlum/etc/template/Enclave.pem
SGX mode: HW
Building new image...
[+] Home dir is /root
[-] Open token file /root/enclave.token error! Will create one.
[+] Init Enclave Successful 15715285336066!
Building libOS...
EXPORT => OCCLUM_BUILTIN_CONF_FILE_MAC = f2-92-39-2e-ce-99-a6-77-de-64-2d-66-24-5a-66-70
EXPORT => OCCLUM_BUILTIN_VM_USER_SPACE_SIZE = 2684354560
Signing the enclave...
<EnclaveConfiguration>
    <ProdID>0</ProdID>
    <ISVSVN>0</ISVSVN>
    <StackMaxSize>10485760</StackMaxSize>
    <HeapMaxSize>335544320</HeapMaxSize>
    <TCSNum>32</TCSNum>
    <TCSPolicy>1</TCSPolicy>
    <DisableDebug>0</DisableDebug>
    <MiscSelect>0</MiscSelect>
    <MiscMask>0xFFFFFFFF</MiscMask>
    <ReservedMemMaxSize>2684354560</ReservedMemMaxSize>
    <ReservedMemMinSize>2684354560</ReservedMemMinSize>
    <ReservedMemInitSize>2684354560</ReservedMemInitSize>
    <ReservedMemExecutable>1</ReservedMemExecutable>
</EnclaveConfiguration>
tcs_num 32, tcs_max_num 32, tcs_min_pool 1
The required memory is 3358101504B.
The required memory is 0xc8289000, 3279396 KB.
Succeed.
Built the Occlum image and enclave successfully
occlum run /bin/occlum_pong
grpc_pingpong server is waiting for service requests ...
Receiving Ping: Hello (at 2020-09-01T04:12:53.538707748Z)

```

client side:
```
root@07ae377ab8cf:~/demos/golang/grpc_pingpong#
root@07ae377ab8cf:~/demos/golang/grpc_pingpong# ./run_ping_on_occlum.sh
/root/demos/golang/grpc_pingpong/occlum_ping_instance initialized as an Occlum instance
Enclave sign-tool: /opt/intel/sgxsdk/bin/x64/sgx_sign
Enclave sign-key: /opt/occlum/etc/template/Enclave.pem
SGX mode: HW
Building new image...
[+] Home dir is /root
[+] Open token file success!
[+] Token file invalid, will create new token file
[+] Init Enclave Successful 16376710299650!
Building libOS...
EXPORT => OCCLUM_BUILTIN_CONF_FILE_MAC = e8-68-a8-e2-1b-c4-1f-45-7d-ba-a9-d5-ee-f1-ba-9b
EXPORT => OCCLUM_BUILTIN_VM_USER_SPACE_SIZE = 2684354560
Signing the enclave...
<EnclaveConfiguration>
    <ProdID>0</ProdID>
    <ISVSVN>0</ISVSVN>
    <StackMaxSize>10485760</StackMaxSize>
    <HeapMaxSize>335544320</HeapMaxSize>
    <TCSNum>32</TCSNum>
    <TCSPolicy>1</TCSPolicy>
    <DisableDebug>0</DisableDebug>
    <MiscSelect>0</MiscSelect>
    <MiscMask>0xFFFFFFFF</MiscMask>
    <ReservedMemMaxSize>2684354560</ReservedMemMaxSize>
    <ReservedMemMinSize>2684354560</ReservedMemMinSize>
    <ReservedMemInitSize>2684354560</ReservedMemInitSize>
    <ReservedMemExecutable>1</ReservedMemExecutable>
</EnclaveConfiguration>
tcs_num 32, tcs_max_num 32, tcs_min_pool 1
The required memory is 3358101504B.
The required memory is 0xc8289000, 3279396 KB.
Succeed.
Built the Occlum image and enclave successfully
occlum run /bin/occlum_ping
Ping: at 2020-09-01T04:12:53.516955702Z
Pong from server: Greetings from Pong! Ping Echoed: Hello (at 2020-09-01T04:12:53.538707748Z)
End-to-End latency is: 0.023362 Seconds

real    0m20.941s
user    0m9.032s
sys     0m11.814s
root@07ae377ab8cf:~/demos/golang/grpc_pingpong#

```